source/common/upstream/load_balancer_impl.cc

```diff
@@ -1299,6 +1299,33 @@ HostConstSharedPtr LeastRequestLoadBalancer::unweightedHostPick(const HostVector
                                                                 const HostsSource&) {
   HostSharedPtr candidate_host = nullptr;
 
+  // Use a full scan if:
+  // - it's explicitly configured, or
+  // - the number of choices is equal to or larger than the number of hosts.
+  const auto has_full_scan_configured = selection_method_ ==
+                                        envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest_SelectionMethod_FULL_SCAN;
+  if (has_full_scan_configured || (hosts_to_use.size() <= choice_count_)) {
+    // Choose a random index to start from preventing always picking the first host in the list.
+    const int rand_idx = random_.random() % hosts_to_use.size();
```

Comment:
This is still going to be problematic as far as selection probabilities go. Consider some host vector with the following weights:

[9, 9, 1, 1, 1]

Choosing a random index to start the scan from would still choose the first host 80% of the time. The host at index 2 will only be picked 20% of the time, which seems unintuitive.

Reply:
Can you walk me through why that happens?

I thought `hosts_to_use.size()` would be 5, thinking it was just a `std::vector`. Your comment makes me think something else is happening, like `hosts_to_use.size()` returns a sum of weights? But presumably the array is still of size 5, so I shouldn't be able to index past [4] in the for-loop that iterates through hosts. I feel like I'm missing something.

Reply:
Ah, that makes sense. If there's a tie for "least requests", then hosts with more non-least requests before them in the array will get more selection weight.

> It can be easily fixed, but it's gonna make the implementation a bit less performant.

I haven't thought about the best way to easily fix this yet. If you have a suggestion in mind, I'm open to it.

Reply:
Here's some pseudo-code:
```
std::vector<HostPtr> candidates;
candidates.reserve(hosts_to_use.size());
for (const auto& host : hosts_to_use) {
  const bool reset_candidates = 
      candidates->empty() || (host.active_rqs < candidates.front()->active_rqs);
  if (reset_candidates) {
    // Found a new minimum active request count.
    candidates.clear();
    candidates.push_back(host);
  } else if (host.active_rqs == candidates.front()->active_rqs) {
    // Found another host with the minimum number of active requests.
    candidates.push_back(host);
  }
}

return candidates[rand_idx % candidates.size()];
```

Basically just find all the hosts that share the smallest number of active requests and choose one of them at random. There's probably some fancy modern std::algorithm function that does this in a more elegant way, but you get the idea.

Reply:
I think I've found an approach that doesn't require allocating a new array. I'll write the code and you can evaluate (including on your test bench) whether it gets acceptably close to the ideal pick ratios. If not, I'll fallback to the straightforward approach you gave above.

Reply:
@tonya11en : Here's the approach I implemented for your consideration. Let me know your thoughts. Can fallback to the original psuedo-code you posed if needed.

https://github.com/envoyproxy/envoy/pull/31507/files#diff-33eefa2339d273434d9a8edee82d672f450eb13c0c956d637f8a6f134d1319afR1314-R1357

Reply:
Actually, I have to change the random start index to be decoupled from the random number used to select the prime. I'll ping you again here when that's done.

Reply:
Done - decoupled the random number used to select a large prime from the random number used to select the pre-increment array index.

Reply:
Keep in mind, this approach will always perform somewhere between the "random start index, iterate by 1" approach and the pseudo-code you wrote above in terms of following the pick ratios. In most cases, I think it will get close enough to the ideal that your pseudo-code achieves. However, there will be some cases where it looks closer to the "random start index, iterate by 1" approach.

The worst case that comes to mind is with 6 hosts that have active request counts of `[9, 9, 9, 1, 1, 1]`. 6 hosts only has two possible increment values (1 and 5 are the only coprimes with 6 that are <6). 6 is the highest count with only 2 possible increment values (that happen to be +1 and -1 increment). In that case, the pick ratios will be:
- Indices 0-2: 0%
- Index 3: 41.66% (5/12)
- Index 4: 16.66% (2/12)
- Index 5: 41.66% (5/12)

The ideal pick ratio is that indices 3-5 will all get 33.33%.

The pick ratios should approach the ideal as the number of possible increment values increases (coprimes with the host count). Here's what the possible increment values look like by host count:
- 1: 0
- 2: 1
- 3: 1, 2
- 4: 1, 3
- 5: 1, 2, 3, 4
- 6: 1, 5
- 7: 1, 2, 3, 4, 5, 6
- 8: 1, 3, 5, 7
- 9: 1, 2, 4, 5, 7, 8
- 10: 1, 3, 7, 9

---

If it's deemed critical that the selection match the ideal pick ratio in all cases, then this approach is not great. However, we could potentially have two different modes in that case... one that is `FULL_SCAN` that will follow the ideal pick ratio in all cases, and one that is `FULL_SCAN_BIASED` with much lower memory overhead with that trade-off that it doesn't always follow the ideal pick ratio (and how much it deviates depends on the host count, how many ties for "least requests" exist, and how those ties are distributed in the host array).